File: frontend/lib/task.ts
import api from "./api";

export async function getTasks() {
  const response = await api.get("/tasks/");
  return response.data;
}

export async function getTask(id: number) {
  const response = await api.get(`/tasks/${id}`);
  return response.data;
}

export async function createTask(data: any) {
  const response = await api.post("/tasks/", data);
  return response.data;
}

export async function updateTask(id: number, data: any) {
  const response = await api.put(`/tasks/${id}`, data);
  return response.data;
}

export async function deleteTask(id: number) {
  const response = await api.delete(`/tasks/${id}`);
  return response.data;
}

File: backend/app/routers/task.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.auth.dependencies import get_current_user
from app.schemas.task import TaskCreate, TaskUpdate
from app.services.task_service import (
    create_task,
    get_tasks,
    get_task,
    update_task,
    delete_task,
)

router = APIRouter(
    prefix="/tasks",
    tags=["Tasks"]
)


@router.post("/")
def add_task(
    task: TaskCreate,
    db: Session = Depends(get_db),
    current_user=Depends(get_current_user)
):
    return create_task(db, task)


@router.get("/")
def all_tasks(
    db: Session = Depends(get_db),
    current_user=Depends(get_current_user)
):
    return get_tasks(db)


@router.get("/{task_id}")
def single_task(
    task_id: int,
    db: Session = Depends(get_db),
    current_user=Depends(get_current_user)
):
    task = get_task(db, task_id)

    if task is None:
        raise HTTPException(status_code=404, detail="Task not found")

    return task


@router.put("/{task_id}")
def edit_task(
    task_id: int,
    task: TaskUpdate,
    db: Session = Depends(get_db),
    current_user=Depends(get_current_user)
):
    updated = update_task(db, task_id, task)

    if updated is None:
        raise HTTPException(status_code=404, detail="Task not found")

    return updated


@router.delete("/{task_id}")
def remove_task(
    task_id: int,
    db: Session = Depends(get_db),
    current_user=Depends(get_current_user)
):
    deleted = delete_task(db, task_id)

    if deleted is None:
        raise HTTPException(status_code=404, detail="Task not found")

    return deleted